### ISSUE TITLE
Feature: Multiple messages per channel

### ISSUE THREAD

**Issue by febus982** [airtai/faststream#1566]
**Is your feature request related to a problem? Please describe.**
I have an existing application ecosystem that produces multiple messages per channel, with different headers per message types.

I have tried this:
```
broker.publisher(topic, schema=Union[EventA, EventB], headers={"aaa": "ccc"})
```

It generated multiple `payloads` in the AsyncAPI json file, but `headers` seem to be ignored (and it would be the same headers for all messages)

**Describe the solution you'd like**
- Be able to register multiple message types per channel
- It should be possible to specify custom headers for each message, as per [OpenAPI](https://www.asyncapi.com/docs/reference/specification/v3.0.0#messageObject) spec

**Feature code example**
Something like this would be great. Maybe using a base message class that defines `get_payload` and `get_headers` methods.

```python
class BaseMessage:
    @classmethod
    def get_headers(cls):
        ...
    @classmethod
    def get_payload(cls):
        ...

class MessageA:
    ...
class MessageB:
    ...

broker.publisher("channel_name", messages: Union[MessageA, MessageB, etc])
...
```

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

**Comment by febus982** [comment]
Just the headers per message would remain.

**Comment by Lancetnik** [comment]
I still can't figure out, what you are trying to do, sorry

**Comment by febus982** [comment]
I was trying to do 2 things.

1. Using a single subscriber for multiple event schemas on a single topic to make sure events are processed in order (e.g. you don't want to process a `BookUpdated` event before a `BookCreated` one) and to have the schema representing this. This I have done with the discriminator in the annotation.

2. I am using CloudEvent specs in my system. The CloudEvent spec defines part of the event metadata (I.e. type, time, other metadata) being sent using headers when using Kafka (and other brokers). Different events have different metadata (I.e. type represents the event class `order-created`/`order-updated`). Doing this would require having the ability of declaring what headers we send for each message type (`OrderCreated` sends `type=order-created`, and so on).

I suspect the functionality in 2 can be achieved by playing a little with custom serializers/decoders but I doubt AsyncApi documentation will reflect any custom serializers/decoders implementation.

I hope this is a bit clearer.

**Comment by Lancetnik** [comment]
@febus982 yeah, desriminator works if you need to process close to each other schemas by the same handler
If you need to use different handlers for different messages, you can just use our [filtering feature](https://faststream.airt.ai/latest/getting-started/subscription/filtering/). Did you check it?

About **AsyncAPI** - unfortunately, our implementation is very limited now and we doesn't support headers representation in it. But we have a plan to improve and polish it to support all **AsyncAPI** features

Can we refactor this issue to make clearer what you want from us? Should we support headers in AsyncAPI or something else? Now I figured out, what you want to do, but still can't undestand, what you want us to do 🥲

**Comment by febus982** [comment]
@Lancetnik Apologies for the late reply. 🙇 

Using different handlers with filtering for different messages would allow to process them, but it would fail to guarantee message ordering. (i.e. in a `user` channel, if you send a `userUpdated` and immediately after a `userDeleted` message, you'll want to make sure the `userUpdated` is processed before the `userDeleted` one).

Answering your question, what I'd like for faststream to have is "support message headers in the OpenAPI spec by reading them from a pydantic class (with support for discriminated unions)"

I would personally try to break this up:
- Define what message class fields are headers (a class method? a list of fields passed to the publisher/subscriber? not sure)
- Validate the message class has the field, probably having it defined as a `Literal` or getting the pydantic `Field` specs as you already do for the body.
- Generate the message headers spec

**Comment by Lancetnik** [comment]
Well, can we summarize features you expect by this Issue? About headers in AsyncAPI - it shouldn't be a problem
We have a special field already

```python
@broker.subscruber(...)
async def handler(header: Annotated[str, Header()]): ...
```

We can use it to generate AsyncAPI schema as well. But should we make something else to support your case?
I am still confused by messages ordering - it is a complex business logic and I am not sure about framework support for it.

**Comment by febus982** [comment]
> Well, can we summarize features you expect by this Issue? About headers in AsyncAPI - it shouldn't be a problem We have a special field already
> 
> ```python
> @broker.subscruber(...)
> async def handler(header: Annotated[str, Header()]): ...
> ```

This is interesting! I see the documentation about what Headers to access, which could be rendered in the AsyncAPI spec.

However we are missing the other scenario (publishing), I think.

> We can use it to generate AsyncAPI schema as well. But should we make something else to support your case? I am still confused by messages ordering - it is a complex business logic and I am not sure about framework support for it.

Strong ordering guarantee is an important capability of stream-based event systems like kafka. 

If we look at the [AsyncAPI spec](https://v2.asyncapi.com/docs/reference/specification/v2.6.0#operationObject) we can see the Headers are defined as part of the Message, not the operation. This is because having different message schemas in the same channel is a valid use case, to support ordering guarantees.

While we can use Discriminated unions to achieve this, what is missing is the capability of defining headers for each message, rather than for operation.

**Comment by febus982** [comment]
One of the problems could be that `pydantic` supports OpenAPI but not AsyncAPI. A quick search led me to https://github.com/asynq-io/pydantic-asyncapi, not sure if providing a base class for messages similar to this could be an option.

**Comment by Lancetnik** [comment]
> ~One of the problems could be that `pydantic` supports OpenAPI but not AsyncAPI for schema generation. A quick search led me to https://github.com/asynq-io/pydantic-asyncapi, not sure if providing a base class for messages similar to this could be an option.~
> 
> I think the above package is only for the schema, but the idea of expanding `pydantic` base model to allow defining the AsyncAPI schema could be a nice way to go (perhaps a Mixin).

I don't think, that we should base our logic on pydantic. It is optional dependency in **FastStream** and can be replaced by **msgspec** or any other serializer as well. So, we should use **FastStream** (or **FastDepends**) native mechanism everywhere

**Comment by Lancetnik** [comment]
> Strong ordering guarantee is an important capability of stream-based event systems like kafka.

Yes, **Kafka**, but not the framework. **FastStream** is just a client for your broker. It reads a messages and publishes them, no more. Such fat abstractions can't be implemented in our ideology

**Comment by febus982** [comment]
> I don't think, that we should base our logic on pydantic. It is optional dependency in FastStream and can be replaced by msgspec or any other serializer as well. So, we should use FastStream (or FastDepends) native mechanism everywhere

Agree

> Yes, Kafka, but not the framework. FastStream is just a client for your broker. It reads a messages and publishes them, no more. Such fat abstractions can't be implemented in our ideology

Agree

> what is missing is the capability of defining headers for each message, rather than for operation.

Just to recap, this requirement is not related to ordering guarantees or pydantic, just AsyncAPI compliance ☝️
